gcsfs/core.py

```diff
@@ -344,32 +373,168 @@ def _call(self, method, path, *args, **kwargs):
+            'get', 'b/{}/o/', bucket, delimiter="/", prefix=prefix, maxResults=max_results)
+
+        assert page["kind"] == "storage#objects"
+        prefixes.extend(page.get("prefixes", []))
+        items.extend(page.get("items", []))
+        next_page_token = page.get('nextPageToken', None)
+
+        while next_page_token is not None:
+            page = self._call(
+                'get', 'b/{}/o/', bucket, delimiter="/", prefix=prefix, maxResults=max_results,
+                pageToken=next_page_token)
+
+            assert page["kind"] == "storage#objects"
+            prefixes.extend(page.get("prefixes", []))
+            items.extend(page.get("items", []))
+            next_page_token = page.get('nextPageToken', None)
+
+        result = {
+            "kind" : "storage#objects",
+            "prefixes" : prefixes,
+            "items" : items,
+        }
+
+        logger.debug("_list_objects result: %s", {k : len(result[k]) for k in ("prefixes", "items")})
+        items = [self._process_object(bucket, i) for i in items]
+
+        return result
+
     def _list_buckets(self):
-        if '' not in self.dirs:
-            try:
-                out = self._call('get', 'b/', project=self.project)
-                dirs = out.get('items', [])
-            except (FileNotFoundError, IOError, ValueError):
-                dirs = []
-            self.dirs[''] = dirs
-        return self.dirs['']
-
-    def _list_bucket(self, bucket, max_results=1000):
-        if bucket not in self.dirs:
-            out = self._call('get', 'b/{}/o/', bucket, maxResults=max_results)
-            dirs = out.get('items', [])
-            next_page_token = out.get('nextPageToken', None)
-            while next_page_token is not None:
-                out = self._call('get', 'b/{}/o/', bucket,
-                                 maxResults=max_results,
-                                 pageToken=next_page_token)
-                dirs.extend(out.get('items', []))
-                next_page_token = out.get('nextPageToken', None)
-            for f in dirs:
-                f['name'] = '%s/%s' % (bucket, f['name'])
-                f['size'] = int(f.get('size'), 0)
-            self.dirs[bucket] = dirs
-        return self.dirs[bucket]
+        """Return list of all buckets under the current project."""
+
+        logger.debug("_list_buckets")
+
+        items = []
+        page = self._call(
+            'get', 'b/', project=self.project
+        )
+
+        assert page["kind"] == "storage#buckets"
+        items.extend(page.get("items", []))
+        next_page_token = page.get('nextPageToken', None)
+
+        while next_page_token is not None:
+            page = self._call(
+                'get', 'b/', project=self.roject, pageToken=next_page_token)
+
+            assert page["kind"] == "storage#buckets"
+            items.extend(page.get("items", []))
+            next_page_token = page.get('nextPageToken', None)
+
+        result = {
```

Comment:
Why the dict? As far as I can see, the only place that this is used, we immediately pick out the 'items' key.

Reply:
This is repacking the result in the form of a de-paginated view of the standard GCS object listing. The `prefixes` list it used later to generated pseudo-directory listings for `ls` and `info` calls.